### ISSUE TITLE
False positive for uninit lint

### ISSUE THREAD

**Issue by zesterer** [rust-lang/rust-clippy#10551]
### Summary

Code like [this](https://github.com/zesterer/chumsky/blob/2be5cd420666e9d6ec07cbf89eb6440be2723d1d/src/private.rs#L277-L281) produces an erroneous lint emission despite the code being safe (`MaybeUninit<T>` has no inhabitants).

### Lint Name

uninit_assumed_init

### Reproducer

I tried this code:

```rust
trait MaybeUninitExt<T>: Sized {
    /// Identical to the unstable [`MaybeUninit::uninit_array`]
    fn uninit_array<const N: usize>() -> [Self; N];
}

impl<T> MaybeUninitExt<T> for MaybeUninit<T> {
    fn uninit_array<const N: usize>() -> [Self; N] {
        // SAFETY: Output type is entirely uninhabited - IE, it's made up entirely of `MaybeUninit`
        unsafe { MaybeUninit::uninit().assume_init() }
    }
}
```

I saw this happen:

```
error: this call for this type may be undefined behavior
   --> src/private.rs:280:18
    |
280 |         unsafe { MaybeUninit::uninit().assume_init() }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
```

I expected to see this happen:

*nothing*

**Comment by zesterer** [comment]
cc @Nilstrieb

**Comment by Nilstrieb** [comment]
See https://github.com/rust-lang/rust-clippy/pull/10520#issuecomment-1484086196 for why this happens

**Comment by Alexendoo** [comment]
Using the old logic as the fallback sounds good to me
